### PR TITLE
Create failed query when session is invalid

### PR DESCRIPTION
**Pull request by dain**
When the server recieves a properly formed Session but the contents are
invalid, create a failed query instead of returning an HTTP error response.
This allows clients to report errors using the normal failed query system.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java [hunk 1]
```diff
@@ -306,6 +328,14 @@ public QueryInfo createQuery(Session session, String query)
             // This is intentionally not a method, since after the state change listener is registered
             // it's not safe to do any of this, and we had bugs before where people reused this code in a method
             URI self = locationFactory.createQueryLocation(queryId);
+
+            // if session creation failed, create a minimal session object
+            if (session == null) {
+                session = Session.builder(new SessionPropertyManager())
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: [Issue]
This seems like a fake session object without full capabilities, but cannot be recognised as such in any way other than context.
Especially using `new SessionPropertyManager()` seems like a source of future problems with this Session object usage. Cannot we use injected sessionPropertyManager instead?

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java [hunk 2]
```diff
@@ -306,6 +328,14 @@ public QueryInfo createQuery(Session session, String query)
...
+                        .setQueryId(queryId)
+                        .setIdentity(sessionFactory.getIdentity())
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: [Issue]
AFAIU This seems like a potential problem with security. It's possible that creating session will fail in access control because of real permissions problem for that identity. And here we will create Session with "validated" user identity.

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+
+    @Override
+    public Session createSession(
+            QueryId queryId,
+            TransactionManager transactionManager,
+            AccessControl accessControl,
+            SessionPropertyManager sessionPropertyManager)
+
+    {
+        accessControl.checkCanSetUser(identity.getPrincipal().orElse(null), identity.getUser());
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java]: [Note]
It's hard to track changes when move of code and changes are both done in same commit.

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+
+        SessionBuilder sessionBuilder = Session.builder(sessionPropertyManager)
+                .setQueryId(queryId)
+                .setIdentity(identity)
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java]: [Note]
It will leak Identity between Sessions if we ever decide to make it mutable. This seems like a bad idea especially that java has no mechanism of forcing immutability and Identity name doesn't suggest that it is immutable by convention.

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+
+    private final Identity identity;
+
+    private final String source;
+    private final String userAgent;
+    private final String remoteUserAddress;
+    private final String timeZoneId;
+    private final String language;
+
+    private final Map<String, String> systemProperties = new HashMap<>();
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java]: [Note]
I think using ImmutableMap would help with catching possible coding errors

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+    private final Map<String, Map<String, String>> catalogSessionProperties = new HashMap<>();
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java]: [Note]
I think using ImmutableMap would help with catching possible coding errors

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+        Splitter splitter = Splitter.on(',').trimResults().omitEmptyStrings();
+        return Collections.list(headers).stream()
+                .map(splitter::splitToList)
+                .flatMap(Collection::stream)
+                .collect(toImmutableList());
+    }
+
+    private static Map<String, String> parseSessionHeaders(HttpServletRequest servletRequest)
+    {
+        Map<String, String> sessionProperties = new HashMap<>();
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java]: [Note]
I think using ImmutableMap would help with catching possible coding errors

File: presto-main/src/main/java/com/facebook/presto/server/ResourceUtil.java [hunk 8]
```diff
@@ -315,7 +191,7 @@ public static String urlEncode(String value)
         }
     }
 
-    public static String urlDecode(String value)
+    private static String urlDecode(String value)
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/server/ResourceUtil.java]: [Note]
Why hide method from Util class and copy it instead of accessing from outside?
Also it's kinda funny that it was public even though it wasn't used outside, but we make it private when we may actually use it from outside ;-) From that perspective it can be interpreted as unrelated change ;-)

If it doesn't fit this Util class I would consider moving that to some new and better place.

**Comment by fiedukow** [comment]
Hey,
Any thoughts on that or plans to merge this or look for other ways of implementing that?
This (as alternative to https://github.com/prestodb/presto/pull/6199) is blocking some other PRs namely: https://github.com/prestodb/presto/pull/5445 which is 6 moths old. Maybe we can wrap it up?

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java (hunk 4 above)
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java]: Value classes like this are always mutable in Presto. If we needed it to be mutable, we'd add a new class like `IdentityBuilder`.

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java [hunk 9]
```diff
@@ -138,24 +126,75 @@ else if (nameParts.size() == 2) {
...
-                sessionBuilder.setCatalogSessionProperty(catalogName, propertyName, propertyValue);
+                catalogSessionProperties.computeIfAbsent(catalogName, id -> new HashMap<>()).put(propertyName, propertyValue);
             }
             else {
                 throw badRequest(format("Invalid %s header", PRESTO_SESSION));
             }
         }
+        this.systemProperties = systemProperties.build();
+        this.catalogSessionProperties = catalogSessionProperties.entrySet().stream()
+                .collect(ImmutableCollectors.toImmutableMap(Entry::getKey, entry -> ImmutableMap.copyOf(entry.getValue())));
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java]: Static import `toImmutableMap `

File: presto-main/src/main/java/com/facebook/presto/server/SessionFactory.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+import com.facebook.presto.spi.QueryId;
+import com.facebook.presto.spi.security.Identity;
+import com.facebook.presto.transaction.TransactionManager;
+
+public interface SessionFactory
+{
+    Identity getIdentity();
+
+    Session createSession(
+            QueryId queryId,
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/SessionFactory.java]: I also had confusion about this when reading the code. After discussing with @dain, the pattern is more clear: this is creating a closure over the request information, so that we can delay creation/validation and thus create a failed query rather than just returning an HTTP error.

The name `SessionFactory` seems bad, though I can't think of a better one.

Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/SessionFactory.java]: While reading the code, I thought I had a simpler alternative, and turns out my idea was the same as what this is already doing.

File: presto-main/src/main/java/com/facebook/presto/server/StatementResource.java [hunk 11]
```diff
@@ -164,12 +151,18 @@ public Response createQuery(
             @Context UriInfo uriInfo)
             throws InterruptedException
     {
-        assertRequest(!isNullOrEmpty(statement), "SQL statement is empty");
+        if (isNullOrEmpty(statement)) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/StatementResource.java]: How about we keep `ResourceUtil` around with the methods `assertRequest`, `urlEncode` and `urlDeocde`?

This class is already long, so I'd rather keep it shorter and avoid duplication.

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java [hunk 12]
```diff
@@ -183,7 +222,8 @@ else if (nameParts.size() == 2) {
         return sessionProperties;
     }
 
-    public static void assertRequest(boolean expression, String format, Object... args)
+    private static void assertRequest(boolean expression, String format, Object... args)
+            throws WebApplicationException
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java]: Not technically needed since this is a `RuntimeException`
